### PR TITLE
Support pointers in un/marshal functions

### PR DESCRIPTION
**Pull request by lwc**
## Background
When updating protobuf to new v2 API we started to see a bunch of linting errors like:
```
web/graph/types/timestamp.go:13:25: MarshalTimestamp passes lock by value: 
google.golang.org/protobuf/types/known/timestamppb.Timestamp 
contains google.golang.org/protobuf/internal/impl.MessageState 
contains sync.Mutex
```

We use [custom scalars for third party types](https://gqlgen.com/reference/scalars/#custom-scalars-with-third-party-types) to bind protobuf timestamps, upon updating the un/marshal functions to act on pointers, I ran into the issue outlined in #1259.

Currently gqlgen assumes that all concrete binding can happen against value types and pointers will delegate through to the underlying value type.

## This PR
Updates binding code to bind to support binding to pointers directly, supporting binding between pointers and values in both directions.

Fixes: #1259 

I have:
 - [ ] Added tests covering the bug / feature (see [testing](https://github.com/99designs/gqlgen/blob/master/TESTING.md))
 - [ ] Updated any relevant documentation (see [docs](https://github.com/99designs/gqlgen/tree/master/docs/content))

### CODE REVIEW

**Comment by lwc** [comment]
This branch has been running 99designs production workloads without issue so I'm inclined to merge it.

**Comment by vektah** [comment]
> It also removes a special case for binding to named types that wrap strings. This special case existed for a long since removed approach to supporting graphql enums, and the interaction between Target and CastType was proving too hard to reason about.

I hate to break it to you but I think this is going to break the prisma guys, judging from [the issue](https://github.com/99designs/gqlgen/issues/595) that came up last time I tried to remove it :joy:

**Comment by lwc** [comment]
> I hate to break it to you but I think this is going to break the prisma guys, judging from the issue that came up last time I tried to remove it 😂

Bummer 😅 

Oh well, after cleaning up my codegen changes it was easy enough to re-instate (I think)

File: codegen/config/binder.go [hunk 1]
```diff
@@ -167,6 +167,7 @@ type TypeReference struct {
 	Definition  *ast.Definition
 	GQL         *ast.Type
 	GO          types.Type
+	Target      types.Type
```
Comment (vektah) [inline codegen/config/binder.go:170]: Could we leave a comment that tries to explain the difference between Go and Type?

The target is the type of the physical field in the struct, and GO is the derived type when you combine it with the pointeryness required by the ast?

File: codegen/config/binder.go [hunk 2]
```diff
@@ -345,16 +349,27 @@ func (b *Binder) TypeReference(schemaType *ast.Type, bindTarget types.Type) (ret
...
 			ref.GO = fun.Type().(*types.Signature).Params().At(0).Type()
 			ref.Marshaler = fun
 			ref.Unmarshaler = types.NewFunc(0, fun.Pkg(), "Unmarshal"+typeName, nil)
-		case hasMethod(obj.Type(), "MarshalGQL") && hasMethod(obj.Type(), "UnmarshalGQL"):
+		} else if hasMethod(obj.Type(), "MarshalGQL") && hasMethod(obj.Type(), "UnmarshalGQL") {
 			ref.GO = obj.Type()
 			ref.IsMarshaler = true
-		default:
+		} else if underlying := basicUnderlying(obj.Type()); def.IsLeafType() && underlying != nil && underlying.Kind() == types.String {
+			// Special case for named types wrapping strings. Used by default enum implementations.
```
Comment (vektah) [inline codegen/config/binder.go]: :+1:

Comment (lwc) [reply codegen/config/binder.go]: Done. TBH I'd still like to remove this case before 1.0.

File: codegen/config/binder.go (hunk 1 above)
Comment (lwc) [reply codegen/config/binder.go]: Done - it's not even that, it's the difference between how to bind to the type, and how to traverse to the type from a field on another type, where sometimes that is due to gql AST requirements.

I wonder if there's an abstraction that represents both pointer juggling _and_ casting sorta like ` “x is satisfiable by y if i do a, b and c”` 🤔

**Pull request merged**